client/multi-currency-setup/tasks/add-currencies-task/index.js

```diff
@@ -0,0 +1,339 @@
+	const enabledCurrencyCodes = Object.keys( enabledCurrencies );
+	const enabledCurrenciesLength = enabledCurrencyCodes.length;
+	const defaultCurrencyCode = defaultCurrency.code;
+
+	// Prevent the enabled currencies and the store currency from displaying
+	// (enabled currencies include the default currency too).
+	const hiddenCurrencies = [ ...enabledCurrencyCodes ];
+
+	// Prefill the selected currency object.
+	const [ activatedCount, setActivatedCount ] = useState( 0 );
+	const [ selectedCurrencies, setSelectedCurrencies ] = useState( {} );
+	useEffect( () => {
+		setSelectedCurrencies(
+			availableCurrencyCodes.reduce( ( acc, value ) => {
+				acc[ value ] = [
+					...recommendedCurrencyCodes,
+					...enabledCurrencyCodes,
+					defaultCurrencyCode,
+				].includes( value );
+				return acc;
+			}, {} )
+		);
+		/* eslint-disable react-hooks/exhaustive-deps */
+	}, [
+		JSON.stringify( availableCurrencyCodes ),
+		JSON.stringify( enabledCurrencyCodes ),
+	] );
+	/* eslint-enable react-hooks/exhaustive-deps */
+
+	// Currency checkbox state change event
+	const handleChange = ( currencyCode, enabled ) => {
+		setSelectedCurrencies( ( previouslyEnabled ) => ( {
+			...previouslyEnabled,
+			[ currencyCode ]: enabled,
+		} ) );
+	};
+
+	// Search component
+	const [ searchText, setSearchText ] = useState( '' );
+	const handleSearchChange = ( event ) => {
+		setSearchText( event.target.value );
+	};
+	const filteredCurrencyCodes = ( ! searchText
+		? availableCurrencyCodes
+		: availableCurrencyCodes.filter( ( code ) => {
+				const { symbol, name } = availableCurrencies[ code ];
+				return (
+					-1 <
+					`${ symbol } ${ code } ${ name }`
+						.toLocaleLowerCase()
+						.indexOf( searchText.toLocaleLowerCase() )
+				);
+		  } )
+	).filter( ( code ) => {
+		// Hide already enabled ones from the search results.
+		return ! hiddenCurrencies.includes( code );
+	} );
+
+	// This state is used for displaying the checked currencies count on the button
+	// and on the description text displayed after adding those checked currencies.
+	useEffect( () => {
+		if ( isActive ) {
+			const activatedCurrenciesCount =
+				Object.values( selectedCurrencies ).filter( Boolean ).length -
+				enabledCurrenciesLength;
+			setActivatedCount(
+				0 < activatedCurrenciesCount ? activatedCurrenciesCount : 0
+			);
+		}
+	}, [ selectedCurrencies, isActive, enabledCurrenciesLength ] );
+
+	const ContinueButton = () => {
+		const checkedCurrencies = useMemo(
+			() =>
+				Object.entries( selectedCurrencies )
+					.map( ( [ currency, enabled ] ) => enabled && currency )
+					.filter( Boolean ),
+			[]
+		);
+
+		const handleContinueClick = useCallback( () => {
+			setStatus( 'pending' );
+			checkedCurrencies.sort();
+			submitEnabledCurrenciesUpdate( checkedCurrencies );
+			setStatus( 'resolved' );
+			setCompleted(
+				{
+					initialCurrencies: enabledCurrencies,
+				},
+				'multi-currency-settings'
+			);
+		}, [ checkedCurrencies ] );
```

Comment:
Using `useMemo` and `useCallback` here has a higher cost in performance than not using them. Memoization has a cost, using it everywhere will be worse than not using it at all.

Declaring `ContinueButton` inside `AddCurrenciesTask` will cause the former to be redefined on every render, you could avoid that declaring it outside or even memoizing the component definition. But we should consider if memoizing a component that doesn't do any heavy calculation worth the cost.

My 2 cents here, we should think twice before using memoization. And only add it after seeing an issue on the profiler or when we know that a heavy calculation will cause slowness. Otherwise it will make the code harder to read, prone to errors and actually less efficient.

ℹ️  [Kent Dodds](https://kentcdodds.com/blog/usememo-and-usecallback) has a good post about it.

Reply:
IIRC, I had problems delivering props into the `ContinueButton` component, and to quickly make it work, I moved it inside the `AddCurrenciesTask` component. I should've returned back and checked that again, thanks for reminding me about that. I'll do the necessary updates.

Reply:
Done in 73cd066.